MapboxCoreNavigation/RouteProgress.swift

```diff
@@ -211,10 +221,55 @@ open class RouteProgress: NSObject {
 
                     congestionTimesPerStep[legIndex].append(stepCongestionValues)
                 }
+
+                if let segmentMaximumSpeedLimits = leg.segmentMaximumSpeedLimits {
+                    guard nextManeuverCoordinateIndex < segmentMaximumSpeedLimits.count else { continue }
+
+                    let stepSegmentMaximumSpeedLimits = Array(segmentMaximumSpeedLimits[maneuverCoordinateIndex..<nextManeuverCoordinateIndex])
+                    maximumSpeedLimitsByStep.append(stepSegmentMaximumSpeedLimits)
+
+                    var totalDistanceTraveled = 0.0
+                    var previousCoordinate = coordinates.first!
+                    var maximumSpeedPerCoordinate: [PositionedSpeedLimit] = []
+
+                    for (index, speedLimit) in stepSegmentMaximumSpeedLimits.enumerated() {
+                        let currentCoordinate = coordinates[index]
+                        let distanceTraveled = previousCoordinate.distance(to: currentCoordinate)
+                        totalDistanceTraveled += distanceTraveled
+                        maximumSpeedPerCoordinate.append((speedLimit, totalDistanceTraveled))
+                        previousCoordinate = currentCoordinate
+                    }
+
+                    maximumSpeedLimitDistanceByStep.append(maximumSpeedPerCoordinate)
+                }
+
+                maneuverCoordinateIndex = nextManeuverCoordinateIndex
             }
 
             congestionTravelTimesSegmentsByStep.append(congestionTravelTimesSegmentsByLeg)
+            positionedSpeedLimitsByStep.append(maximumSpeedLimitDistanceByStep)
+        }
+    }
+
+    /**
+     Returns the SpeedLimit for the current position along the route. Returns SpeedLimit.invalid if the speed limit is unknown or missing.
+     
+     The maximum speed may be an advisory speed limit for segments where legal limits are not posted, such as highway entrance and exit ramps. If the speed limit along a particular segment is unknown, it is set to `nil`. If the speed is unregulated along the segment, such as on the German _Autobahn_ system, it is represented by a measurement whose value is `Double.greatestFiniteMagnitude`.
+     
+     Speed limit data is available in [a number of countries and territories worldwide](https://docs.mapbox.com/help/how-mapbox-works/directions/).
+     */
+    public var currentSpeedLimit: Measurement<UnitSpeed>? {
+        guard legIndex < positionedSpeedLimitsByStep.count,
+            currentLegProgress.stepIndex < positionedSpeedLimitsByStep[legIndex].count else {
+            return nil
+        }
+        let speedLimits = positionedSpeedLimitsByStep[legIndex][currentLegProgress.stepIndex]
+        let lastSpeedLimitTuple = speedLimits.last { $0.1 <= currentLegProgress.currentStepProgress.distanceTraveled }
```

Comment:
As with any annotation type vended by the Directions API, each annotation corresponds to the span between two coordinates in the route geometry. So instead of scanning the whole step geometry for a matching position, we can get the index of the closest coordinate on the route line using Turf, then index into the array of speed limits. That will eliminate any unreliability caused by floating point imprecision.